File: lib/nodejs/opentoken/src/processor/PersonAttributesProcessor.ts
/**
 * Copyright (c) Truveta. All rights reserved.
 */

import { PersonAttributesReader } from '../io/PersonAttributesReader';
import { PersonAttributesWriter } from '../io/PersonAttributesWriter';
import { TokenTransformer } from '../tokentransformer/TokenTransformer';
import { AttributeLoader } from '../attributes/AttributeLoader';
import { Metadata } from '../Metadata';
import { TokenGenerator } from '../tokens/TokenGenerator';
import { TokenDefinition } from '../tokens/TokenDefinition';
import { Attribute } from '../attributes/Attribute';

/**
 * Process all person attributes.
 *
 * This class is used to read person attributes from input source,
 * generate tokens for each person record and write the tokens back
 * to the output data source.
 */
export class PersonAttributesProcessor {
  /**
   * Reads person attributes from the input data source, generates tokens,
   * and writes the result back to the output data source.
   */
  static async process(
    reader: PersonAttributesReader,
    writer: PersonAttributesWriter,
    transformers: TokenTransformer[],
    metadata: Metadata
  ): Promise<void> {
    const attributes = AttributeLoader.load();
    const tokenDefinition = new TokenDefinition();
    const tokenGenerator = new TokenGenerator(tokenDefinition, transformers);
    let rowCounter = 0;

    // Track invalid attributes and blank tokens
    const invalidAttributeCount = new Map<string, number>();
    const blankTokensByRuleCount = new Map<string, number>();

    while (await reader.hasNext()) {
      const row = await reader.next();
      rowCounter++;

      // Convert row data to person attributes map
      const personAttributes = new Map<new () => Attribute, string>();
      
      for (const attr of attributes) {
        const attrName = attr.getName();
        let value = row.get(attrName);

        // Also try aliases
        if (!value) {
          for (const alias of attr.getAliases()) {
            value = row.get(alias);
            if (value) break;
          }
        }

        if (value) {
          personAttributes.set(attr.constructor as new () => Attribute, value);
        }
      }

      // Generate all tokens for this row
      const result = await tokenGenerator.getAllTokens(personAttributes);

      // Track invalid attributes
      if (result.getInvalidAttributes().size > 0) {
        for (const attrName of result.getInvalidAttributes()) {
          invalidAttributeCount.set(
            attrName,
            (invalidAttributeCount.get(attrName) || 0) + 1
          );
        }
      }

      // Track blank tokens
      if (result.getBlankTokensByRule().size > 0) {
        for (const ruleId of result.getBlankTokensByRule()) {
          blankTokensByRuleCount.set(
            ruleId,
            (blankTokensByRuleCount.get(ruleId) || 0) + 1
          );
        }
      }

      // Write tokens
      const recordId = row.get('RecordId') || row.get('record-id') || `row-${rowCounter}`;
      for (const [ruleId, token] of result.getTokens().entries()) {
        const outputRecord = new Map<string, string>();
        outputRecord.set('RecordId', recordId);
        outputRecord.set('RuleId', ruleId);
        outputRecord.set('Token', token);
        await writer.writeAttributes(outputRecord);
      }

      if (rowCounter % 1000 === 0) {
        console.log(`Processed ${rowCounter} records`);
      }
    }

    // Log invalid attribute statistics
    if (invalidAttributeCount.size > 0) {
      for (const [attrName, count] of invalidAttributeCount.entries()) {
        console.log(
          `Total invalid Attribute count for [${attrName}]: ${count.toLocaleString()}`
        );
      }
      const totalInvalid = Array.from(invalidAttributeCount.values()).reduce(
        (sum, count) => sum + count,
        0
      );
      console.log(
        `Total number of records with invalid attributes: ${totalInvalid.toLocaleString()}`
      );
    }

    // Log blank token statistics
    if (blankTokensByRuleCount.size > 0) {
      for (const [ruleId, count] of blankTokensByRuleCount.entries()) {
        console.log(
          `Total blank tokens for rule [${ruleId}]: ${count.toLocaleString()}`
        );
      }
      const totalBlankTokens = Array.from(
        blankTokensByRuleCount.values()
      ).reduce((sum, count) => sum + count, 0);
      console.log(
        `Total blank tokens generated: ${totalBlankTokens.toLocaleString()}`
      );
    }

    metadata.totalRows = rowCounter;
    console.log(`Processed a total of ${rowCounter} records`);
  }
}

File: lib/java/opentoken/src/main/java/com/truveta/opentoken/processor/PersonAttributesProcessor.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.processor;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;
import java.util.UUID;
import java.util.stream.Collectors;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.truveta.opentoken.attributes.Attribute;
import com.truveta.opentoken.attributes.general.RecordIdAttribute;
import com.truveta.opentoken.io.PersonAttributesReader;
import com.truveta.opentoken.io.PersonAttributesWriter;
import com.truveta.opentoken.tokens.TokenDefinition;
import com.truveta.opentoken.tokens.TokenGenerator;
import com.truveta.opentoken.tokens.TokenGeneratorResult;
import com.truveta.opentoken.tokentransformer.TokenTransformer;

/**
 * Process all person attributes.
 * <p>
 * This class is used to read person attributes from input source,
 * generate tokens for each person record and write the tokens back
 * to the output data source.
 */
public final class PersonAttributesProcessor {

    public static final String TOTAL_ROWS = "TotalRows";
    public static final String TOTAL_ROWS_WITH_INVALID_ATTRIBUTES = "TotalRowsWithInvalidAttributes";
    public static final String INVALID_ATTRIBUTES_BY_TYPE = "InvalidAttributesByType";
    public static final String BLANK_TOKENS_BY_RULE = "BlankTokensByRule";

    private static final Logger logger = LoggerFactory.getLogger(PersonAttributesProcessor.class);

    PersonAttributesProcessor() {
    }

    /**
     * Reads person attributes from the input data source, generates token, and
     * write the result back to the output data source. The tokens can be optionally
     * transformed before writing.
     * 
     * @param reader               the reader initialized with the input data
     *                             source.
     * @param writer               the writer initialized with the output data
     *                             source.
     * @param tokenTransformerList a list of token transformers.
     * 
     * @see com.truveta.opentoken.io.PersonAttributesReader PersonAttributesReader
     * @see com.truveta.opentoken.io.PersonAttributesWriter PersonAttributesWriter
     * @see com.truveta.opentoken.tokentransformer.TokenTransformer TokenTransformer
     */
    public static void process(PersonAttributesReader reader, PersonAttributesWriter writer,
            List<TokenTransformer> tokenTransformerList, Map<String, Object> metadataMap) throws IOException {

        // TokenGenerator code
        TokenDefinition tokenDefinition = new TokenDefinition();
        TokenGenerator tokenGenerator = new TokenGenerator(tokenDefinition, tokenTransformerList);

        Map<Class<? extends Attribute>, String> row;
        TokenGeneratorResult tokenGeneratorResult;

        long rowCounter = 0;
        Map<String, Long> invalidAttributeCount = initializeInvalidAttributeCount();
        Map<String, Long> blankTokensByRuleCount = initializeBlankTokensByRuleCount(tokenDefinition);

        while (reader.hasNext()) {
            row = reader.next();
            rowCounter++;

            tokenGeneratorResult = tokenGenerator.getAllTokens(row);
            logger.debug("Tokens: {}", tokenGeneratorResult.getTokens());

            keepTrackOfInvalidAttributes(tokenGeneratorResult, rowCounter,
                    invalidAttributeCount);

            keepTrackOfBlankTokens(tokenGeneratorResult, rowCounter,
                    blankTokensByRuleCount);

            writeTokens(writer, row, rowCounter, tokenGeneratorResult);

            if (rowCounter % 10000 == 0) {
                logger.info(String.format("Processed \"%,d\" records", rowCounter));
            }
        }

        logger.info(String.format("Processed a total of %,d records", rowCounter));

        invalidAttributeCount
                .forEach((key, value) -> logger
                        .info(String.format("Total invalid Attribute count for [%s]: %,d", key, value)));
        long rowIssueCounter = invalidAttributeCount.values().stream()
                .collect(Collectors.summarizingLong(Long::longValue)).getSum();

        metadataMap.put(TOTAL_ROWS, rowCounter);
        metadataMap.put(TOTAL_ROWS_WITH_INVALID_ATTRIBUTES, rowIssueCounter);
        // Alphabetize attribute and token rule keys for deterministic metadata output
        metadataMap.put(INVALID_ATTRIBUTES_BY_TYPE, new java.util.TreeMap<>(invalidAttributeCount));
        metadataMap.put(BLANK_TOKENS_BY_RULE, new java.util.TreeMap<>(blankTokensByRuleCount));
        logger.info(String.format("Total number of records with invalid attributes: %,d", rowIssueCounter));

        blankTokensByRuleCount
                .forEach((key, value) -> logger
                        .info(String.format("Total blank tokens for rule [%s]: %,d", key, value)));
        long blankTokensTotal = blankTokensByRuleCount.values().stream()
                .collect(Collectors.summarizingLong(Long::longValue)).getSum();
        logger.info(String.format("Total blank tokens generated: %,d", blankTokensTotal));
    }

    private static void writeTokens(PersonAttributesWriter writer, Map<Class<? extends Attribute>, String> row,
            long rowCounter, TokenGeneratorResult tokenGeneratorResult) {

        Set<String> tokenIds = new TreeSet<>(tokenGeneratorResult.getTokens().keySet());

        // Generate a UUID for RecordId if it's not present in the input data
        String recordId = row.get(RecordIdAttribute.class);
        if (recordId == null || recordId.isEmpty()) {
            recordId = UUID.randomUUID().toString();
        }

        for (String tokenId : tokenIds) {
            var rowResult = new HashMap<String, String>();
            rowResult.put(TokenConstants.RECORD_ID, recordId);
            rowResult.put(TokenConstants.RULE_ID, tokenId);
            rowResult.put(TokenConstants.TOKEN, tokenGeneratorResult.getTokens().get(tokenId));

            try {
                writer.writeAttributes(rowResult);
            } catch (IOException e) {
                logger.error(String.format("Error writing attributes to file for row %,d", rowCounter), e);
            }
        }
    }

    private static void keepTrackOfInvalidAttributes(TokenGeneratorResult tokenGeneratorResult, long rowCounter,
            Map<String, Long> invalidAttributeCount) {

        if (!tokenGeneratorResult.getInvalidAttributes().isEmpty()) {
            logger.info("Invalid Attributes for row {}: {}", String.format("%,d", rowCounter),
                    tokenGeneratorResult.getInvalidAttributes());

            for (String invalidAttribute : tokenGeneratorResult.getInvalidAttributes()) {
                if (invalidAttributeCount.containsKey(invalidAttribute)) {
                    invalidAttributeCount.put(invalidAttribute, invalidAttributeCount.get(invalidAttribute) + 1);
                } else {
                    invalidAttributeCount.put(invalidAttribute, 1L);
                }
            }
        }
    }

    private static void keepTrackOfBlankTokens(TokenGeneratorResult tokenGeneratorResult, long rowCounter,
            Map<String, Long> blankTokensByRuleCount) {

        if (!tokenGeneratorResult.getBlankTokensByRule().isEmpty()) {
            logger.debug("Blank tokens for row {}: {}", String.format("%,d", rowCounter),
                    tokenGeneratorResult.getBlankTokensByRule());

            for (String ruleId : tokenGeneratorResult.getBlankTokensByRule()) {
                blankTokensByRuleCount.merge(ruleId, 1L, Long::sum);
            }
        }
    }

    /**
     * Initialize the invalid attribute count map with all registered attributes set to 0.
     * This ensures that all attribute types appear in the metadata even in happy path scenarios.
     *
     * @return a map with all attribute names initialized to 0
     */
    private static Map<String, Long> initializeInvalidAttributeCount() {
        Map<String, Long> invalidAttributeCount = new HashMap<>();
        Set<Attribute> attributes = com.truveta.opentoken.attributes.AttributeLoader.load();
        for (Attribute attribute : attributes) {
            invalidAttributeCount.put(attribute.getName(), 0L);
        }
        return invalidAttributeCount;
    }

    /**
     * Initialize the blank tokens by rule count map with all token identifiers set to 0.
     * This ensures that all token rules appear in the metadata even in happy path scenarios.
     *
     * @param tokenDefinition the token definition containing all token identifiers
     * @return a map with all token identifiers initialized to 0
     */
    private static Map<String, Long> initializeBlankTokensByRuleCount(TokenDefinition tokenDefinition) {
        Map<String, Long> blankTokensByRuleCount = new HashMap<>();
        for (String tokenId : tokenDefinition.getTokenIdentifiers()) {
            blankTokensByRuleCount.put(tokenId, 0L);
        }
        return blankTokensByRuleCount;
    }
}